### DISCUSSION TITLE
[HW2] Problem 2.1

### DISCUSSION THREAD

**Discussion by emjay73** [seominjoon/kaist-ai605#98] [Q&A] [answered]
Problem 2.1 (2 points) We will first start with the Bag of Words that we discussed in Lecture 08. Using the definition in the class (don't worry about the exact definition though), implement BagOfWords class that subclasses SimilaritySearch class.

Even after removing tokens with frequency<2, vocab size is too big so that BoW(n_val_docs x n_vocs) can not be loaded on Colab.
Am I doing something wrong? or do I have to remove less duplicated vocabs more?

And do I need a UNK token for this assignment too?

**Comment by dinhkt** [comment]
Are you using squad_dataset['train']? Professor said that we only use the squad_dataset['validation'] in the first part of the assignment. For me, I can create BoW in colab with no problem and I don't even remove the tokens that frequency <2.

**Reply by emjay73** [reply]
"we will view them as retrieval task by trying to find the correct document corresponding to the question among all the documents in the validation data."
I thought we should make vocab using the training set just we've done in assignment 1 and find documents in the validation set using a question in the validation set. 
So what you suggest is to make vocab using validation set documents and find documents corresponds to question in the validation set?

**Reply by dinhkt** [reply]
yes, I only use the validation set to create vocab. Then I use the questions in the validation set and try to find the most similar documents in the validation set. But it's better to wait for professor answer to make it more clear :)

**Reply by emjay73** [reply]
OK, thank you for sharing your experience though,

**Comment by seominjoon** [comment, marked as answer]
Hi, you do not want to create the full dense vector. It is very inefficient! Rather, you want to store it in an efficient way, e.g. using a dictionary. For example, something like this:
```python
document = "hello world"
vocab = ['hello', 'world'] 
d = {0: 1.0, 1: 1.0} # this is the BoW vector corresponding to the document
```

If this was tf-idf, you will have a value other than 1.0. 
You can also efficiently compute the inner product using this format.

**Reply by emjay73** [reply]
Oh, I see. Thanks.
And about  <PAD> and <UNK>, why do we need <PAD> in this assignment.
Also, is it appropriate to use <UNK> for the inner product?

**Reply by seominjoon** [reply]
I didn't understand your first question.
The beauty of inner product is that you only need to look at the terms in your query (if they are not, inner product for that dimension will be zero anyways).

**Reply by emjay73** [reply]
I used <> symbol in the sentence and it automatically deletes the word in it.
Let me rewrite the question again.
```
And about 'PAD' and 'UNK', why do we need 'PAD' in this assignment.
Also, is it appropriate to use 'UNK'  for the inner product?
```
Thank you.

**Reply by seominjoon** [reply]
Yeah you don't need PAD for tf-idf. I just put it so that you can use the same vocab strcutrue as in Assignment 1, but you are right that PAD is useless here.
You don't want to use UNK for tf-idf; it will have bad side effects. So I think my code was misleading. I am correcting it now.

**Reply by emjay73** [reply]
Thank you!
